File: vscode-extension/src/service/ContextService.ts
/**
 * 上下文服务
 */

import * as vscode from 'vscode';

/**
 * 上下文信息
 */
export interface Context {
    files: FileContext[];
    cursor: CursorPosition;
    selectedCode: string | null;
}

export interface FileContext {
    path: string;
    language: string;
    content: string;
}

export interface CursorPosition {
    line: number;
    column: number;
}

/**
 * 上下文服务
 */
export class ContextService {
    
    /**
     * 提取当前文件上下文
     */
    extractCurrentFile(): FileContext | null {
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            return null;
        }
        
        const document = editor.document;
        
        return {
            path: document.uri.fsPath,
            language: this.detectLanguage(document.languageId),
            content: document.getText(),
        };
    }
    
    /**
     * 提取光标位置
     */
    extractCursor(): CursorPosition {
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            return { line: 1, column: 1 };
        }
        
        const position = editor.selection.active;
        return {
            line: position.line + 1,
            column: position.character + 1,
        };
    }
    
    /**
     * 提取选中的代码
     */
    extractSelection(): string | null {
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            return null;
        }
        
        const selection = editor.selection;
        if (selection.isEmpty) {
            return null;
        }
        
        return editor.document.getText(selection);
    }
    
    /**
     * 构建完整上下文
     */
    buildContext(): Context {
        const files: FileContext[] = [];
        
        const currentFile = this.extractCurrentFile();
        if (currentFile) {
            files.push(currentFile);
        }
        
        return {
            files,
            cursor: this.extractCursor(),
            selectedCode: this.extractSelection(),
        };
    }
    
    /**
     * 检测语言
     */
    private detectLanguage(languageId: string): string {
        const languageMap: Record<string, string> = {
            java: 'java',
            kotlin: 'kotlin',
            python: 'python',
            javascript: 'javascript',
            typescript: 'typescript',
            rust: 'rust',
            go: 'go',
            'c#': 'csharp',
            cpp: 'cpp',
            html: 'html',
            css: 'css',
            json: 'json',
            xml: 'xml',
            markdown: 'markdown',
            sql: 'sql',
        };
        
        return languageMap[languageId.toLowerCase()] ?? 'plaintext';
    }
}

File: idea-plugin/bin/main/com/vex/coding/service/ContextService.kt
package com.vex.coding.service

import com.intellij.openapi.editor.Caret
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager

/**
 * 上下文信息
 */
data class Context(
    val files: List<FileContext>,
    val cursor: CursorPosition,
    val selectedCode: String?
)

data class FileContext(
    val path: String,
    val language: String,
    val content: String
)

data class CursorPosition(
    val line: Int,
    val column: Int
)

/**
 * 上下文提取服务
 */
class ContextService(private val project: Project) {
    
    /**
     * 提取当前文件的上下文
     */
    fun extractCurrentFile(): FileContext? {
        val editor = com.intellij.openapi.fileEditor.FileEditorManager.getInstance(project)
            .selectedTextEditor ?: return null
        
        val document = editor.document ?: return null
        val file = com.intellij.openapi.fileEditor.FileEditorManager.getInstance(project)
            .selectedFiles.firstOrNull() ?: return null
        
        return FileContext(
            path = file.path,
            language = detectLanguage(file),
            content = document.text
        )
    }
    
    /**
     * 提取光标位置
     */
    fun extractCursor(): CursorPosition {
        val editor = com.intellij.openapi.fileEditor.FileEditorManager.getInstance(project)
            .selectedTextEditor ?: return CursorPosition(1, 1)
        
        val caret = editor.caretModel.primaryCaret
        return CursorPosition(
            line = caret.logicalPosition.line + 1,
            column = caret.logicalPosition.column + 1
        )
    }
    
    /**
     * 提取选中的代码
     */
    fun extractSelection(): String? {
        val editor = com.intellij.openapi.fileEditor.FileEditorManager.getInstance(project)
            .selectedTextEditor ?: return null
        
        val selectedText = editor.selectionModel.selectedText
        return selectedText?.takeIf { it.isNotBlank() }
    }
    
    /**
     * 构建完整上下文
     */
    fun buildContext(): Context {
        val files = mutableListOf<FileContext>()
        
        extractCurrentFile()?.let { files.add(it) }
        
        return Context(
            files = files,
            cursor = extractCursor(),
            selectedCode = extractSelection()
        )
    }
    
    /**
     * 检测语言
     */
    private fun detectLanguage(file: VirtualFile): String {
        val ext = file.extension?.lowercase() ?: return "plaintext"
        
        return when (ext) {
            "java" -> "java"
            "kt", "kts" -> "kotlin"
            "py" -> "python"
            "js", "jsx" -> "javascript"
            "ts", "tsx" -> "typescript"
            "rs" -> "rust"
            "go" -> "go"
            "xml" -> "xml"
            "json" -> "json"
            "md" -> "markdown"
            "sql" -> "sql"
            "gradle", "kts" -> "gradle"
            else -> "plaintext"
        }
    }
}